File: lib/presentation/screens/onboarding_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../providers/onboarding_provider.dart';
import '../widgets/permission_card.dart';
import '../widgets/folder_discovery_progress.dart';
import '../../domain/models/onboarding_state.dart';

/// Single-screen onboarding flow with SAF permission picker and folder discovery
/// Per user decision: no multi-step wizard, tabs, or pages
class OnboardingScreen extends ConsumerWidget {
  const OnboardingScreen({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final state = ref.watch(onboardingProvider);

    return Scaffold(
      body: SafeArea(
        child: AnimatedSwitcher(
          duration: const Duration(milliseconds: 300),
          transitionBuilder: (Widget child, Animation<double> animation) {
            return FadeTransition(
              opacity: animation,
              child: SlideTransition(
                position: Tween<Offset>(
                  begin: const Offset(0.0, 0.05),
                  end: Offset.zero,
                ).animate(CurvedAnimation(
                  parent: animation,
                  curve: Curves.easeOutCubic,
                )),
                child: child,
              ),
            );
          },
          child: _buildContent(context, ref, state),
        ),
      ),
    );
  }

  Widget _buildContent(BuildContext context, WidgetRef ref, OnboardingState state) {
    return Padding(
      padding: const EdgeInsets.all(24.0),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          const Spacer(flex: 2),

          // App logo and value proposition
          _buildHeader(context),

          const Spacer(flex: 3),

          // Stage-specific content
          switch (state.stage) {
            OnboardingStage.welcome => _buildWelcomeStage(context, ref),
            OnboardingStage.requestingPermission => _buildRequestingPermissionStage(context),
            OnboardingStage.discovering => _buildDiscoveringStage(context, state),
            OnboardingStage.complete => _buildCompleteStage(context, ref),
            OnboardingStage.error => _buildErrorStage(context, ref, state),
          },

          const Spacer(),
        ],
      ),
    );
  }

  Widget _buildHeader(BuildContext context) {
    return Column(
      children: [
        // App icon/logo
        Container(
          padding: const EdgeInsets.all(20),
          decoration: BoxDecoration(
            color: Theme.of(context).colorScheme.primaryContainer,
            borderRadius: BorderRadius.circular(30),
          ),
          child: Icon(
            Icons.photo_library,
            size: 64,
            color: Theme.of(context).colorScheme.onPrimaryContainer,
          ),
        ),
        const SizedBox(height: 24),

        // Title
        Text(
          'Photo Classifier',
          style: Theme.of(context).textTheme.headlineMedium?.copyWith(
                fontWeight: FontWeight.bold,
              ),
          textAlign: TextAlign.center,
        ),
        const SizedBox(height: 12),

        // Value proposition description
        Text(
          'Automatically organize your photos using on-device machine learning. '
          'Your photos stay private and never leave your device.',
          style: Theme.of(context).textTheme.bodyLarge?.copyWith(
                height: 1.5,
              ),
          textAlign: TextAlign.center,
        ),
      ],
    );
  }

  Widget _buildWelcomeStage(BuildContext context, WidgetRef ref) {
    return Column(
      children: [
        // Permission request card with context explanation
        PermissionCard(
          onGrantPressed: () {
            ref.read(onboardingProvider.notifier).requestPermission();
          },
        ),
        const SizedBox(height: 16),

        // Privacy note
        Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Icon(
              Icons.security,
              size: 16,
              color: Theme.of(context).colorScheme.outline,
            ),
            const SizedBox(width: 8),
            Text(
              'No internet connection required - works offline',
              style: Theme.of(context).textTheme.labelSmall?.copyWith(
                    color: Theme.of(context).colorScheme.outline,
                  ),
            ),
          ],
        ),
      ],
    );
  }

  Widget _buildRequestingPermissionStage(BuildContext context) {
    // Spinner while waiting for permission response (per user decision)
    return Column(
      children: [
        SizedBox(
          width: 64,
          height: 64,
          child: CircularProgressIndicator(
            color: Theme.of(context).colorScheme.primary,
            strokeWidth: 4,
          ),
        ),
        const SizedBox(height: 24),
        Text(
          'Opening folder picker...',
          style: Theme.of(context).textTheme.bodyLarge,
          textAlign: TextAlign.center,
        ),
        const SizedBox(height: 8),
        Text(
          'Select your Pictures folder to continue',
          style: Theme.of(context).textTheme.bodyMedium?.copyWith(
                color: Theme.of(context).hintColor,
              ),
          textAlign: TextAlign.center,
        ),
      ],
    );
  }

  Widget _buildDiscoveringStage(BuildContext context, OnboardingState state) {
    // Show folder discovery progress
    return FolderDiscoveryProgress(
      discoveredFolders: state.discoveredFolders,
      totalPhotos: state.totalPhotos > 0 ? state.totalPhotos : null,
      currentFolder: state.currentFolder,
      isComplete: false,
    );
  }

  Widget _buildCompleteStage(BuildContext context, WidgetRef ref) {
    // Success confirmation before main screen (per user decision)
    final state = ref.watch(onboardingProvider);

    // Auto-navigate after short delay
    Future.delayed(const Duration(milliseconds: 1500), () async {
      await ref.read(onboardingProvider.notifier).completeOnboarding();
      // Navigation handled by listener in main.dart or here
    });

    return Column(
      children: [
        // Success indicator
        Container(
          padding: const EdgeInsets.all(20),
          decoration: BoxDecoration(
            color: Theme.of(context).colorScheme.primaryContainer,
            shape: BoxShape.circle,
          ),
          child: Icon(
            Icons.check_circle,
            size: 64,
            color: Theme.of(context).colorScheme.onPrimaryContainer,
          ),
        ),
        const SizedBox(height: 24),

        // Success message
        Text(
          'Ready to organize!',
          style: Theme.of(context).textTheme.headlineSmall?.copyWith(
                fontWeight: FontWeight.bold,
              ),
          textAlign: TextAlign.center,
        ),
        const SizedBox(height: 12),

        Text(
          'Found ${state.discoveredFolders} folders with ${state.totalPhotos} photos',
          style: Theme.of(context).textTheme.bodyLarge,
          textAlign: TextAlign.center,
        ),
        const SizedBox(height: 32),

        // Manual continue button (in case auto-nav doesn't trigger)
        FilledButton.icon(
          onPressed: () async {
            await ref.read(onboardingProvider.notifier).completeOnboarding();
            if (context.mounted) {
              Navigator.of(context).pushReplacementNamed('/main');
            }
          },
          icon: const Icon(Icons.arrow_forward),
          label: const Text('Continue'),
        ),
      ],
    );
  }

  Widget _buildErrorStage(BuildContext context, WidgetRef ref, OnboardingState state) {
    return Column(
      children: [
        // Error icon
        Container(
          padding: const EdgeInsets.all(20),
          decoration: BoxDecoration(
            color: Theme.of(context).colorScheme.errorContainer,
            shape: BoxShape.circle,
          ),
          child: Icon(
            Icons.error_outline,
            size: 64,
            color: Theme.of(context).colorScheme.onErrorContainer,
          ),
        ),
        const SizedBox(height: 24),

        // Error title
        Text(
          'Something went wrong',
          style: Theme.of(context).textTheme.headlineSmall?.copyWith(
                fontWeight: FontWeight.bold,
              ),
          textAlign: TextAlign.center,
        ),
        const SizedBox(height: 12),

        // Error message
        if (state.errorMessage != null)
          Text(
            state.errorMessage!,
            style: Theme.of(context).textTheme.bodyMedium,
            textAlign: TextAlign.center,
          ),
        const SizedBox(height: 32),

        // Retry button
        FilledButton.icon(
          onPressed: () {
            ref.read(onboardingProvider.notifier).retry();
          },
          icon: const Icon(Icons.refresh),
          label: const Text('Try Again'),
        ),
      ],
    );
  }
}

File: app/src/main/java/com/example/photoorganizer/ui/onboarding/OnboardingScreen.kt
package com.example.photoorganizer.ui.onboarding

import android.app.Activity
import android.net.Uri
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideInVertically
import androidx.compose.animation.slideOutVertically
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Error
import androidx.compose.material.icons.filled.PhotoAlbum
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import com.example.photoorganizer.ui.components.PermissionCard
import kotlinx.coroutines.delay

/**
 * Single-screen onboarding flow per user decision.
 * 
 * Shows app value proposition, requests SAF permission with context,
 * displays folder discovery progress, and completes to main screen.
 * No multi-step wizard - everything on one screen.
 *
 * @param onNavigateToMain Callback when onboarding completes
 * @param viewModel The onboarding ViewModel (injected via Hilt)
 */
@Composable
fun OnboardingScreen(
    onNavigateToMain: () -> Unit,
    viewModel: OnboardingViewModel = hiltViewModel()
) {
    val uiState by viewModel.uiState.collectAsState()
    val snackbarHostState = remember { SnackbarHostState() }
    val context = LocalContext.current

    // Launcher for SAF document tree picker
    val launcher = rememberLauncherForActivityResult(
        contract = ActivityResultContracts.OpenDocumentTree()
    ) { uri: Uri? ->
        if (uri != null) {
            // Persist permission immediately
            context.contentResolver.takePersistableUriPermission(
                uri,
                android.content.Intent.FLAG_GRANT_READ_URI_PERMISSION or
                        android.content.Intent.FLAG_GRANT_WRITE_URI_PERMISSION
            )
            viewModel.onPermissionGranted(uri)
        } else {
            viewModel.onPermissionDenied()
        }
    }

    // Handle navigation when onboarding completes
    LaunchedEffect(uiState.stage) {
        if (uiState.stage == OnboardingStage.COMPLETE) {
            delay(1500) // Brief pause to show success state
            onNavigateToMain()
        }
    }

    // Show error in snackbar
    LaunchedEffect(uiState.errorMessage) {
        uiState.errorMessage?.let { message ->
            snackbarHostState.showSnackbar(message)
            viewModel.clearError()
        }
    }

    Scaffold(
        snackbarHost = { SnackbarHost(snackbarHostState) }
    ) { padding ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(padding)
                .padding(horizontal = 24.dp, vertical = 32.dp),
            contentAlignment = Alignment.Center
        ) {
            AnimatedContent(
                targetState = uiState.stage,
                transitionSpec = {
                    fadeIn() + slideInVertically { it / 4 } togetherWith
                            fadeOut() + slideOutVertically { -it / 4 }
                },
                label = "onboarding_content"
            ) { stage ->
                when (stage) {
                    OnboardingStage.WELCOME,
                    OnboardingStage.REQUESTING_PERMISSION -> WelcomeContent(
                        isLoading = uiState.isLoading,
                        onGrantClick = {
                            viewModel.onRequestPermissionClick()
                            launcher.launch(null)
                        }
                    )

                    OnboardingStage.DISCOVERING -> DiscoveryContent(
                        discoveredFolders = uiState.discoveredFolders,
                        totalPhotos = uiState.totalPhotos,
                        isIndeterminate = true
                    )

                    OnboardingStage.COMPLETE -> CompleteContent(
                        discoveredFolders = uiState.discoveredFolders,
                        totalPhotos = uiState.totalPhotos
                    )

                    OnboardingStage.ERROR -> ErrorContent(
                        errorMessage = uiState.errorMessage ?: "An error occurred",
                        onRetry = { viewModel.onRetry() }
                    )
                }
            }
        }
    }
}

/**
 * Welcome content showing app value proposition and permission request.
 */
@Composable
private fun WelcomeContent(
    isLoading: Boolean,
    onGrantClick: () -> Unit
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState()),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(24.dp)
    ) {
        // App icon/logo
        Icon(
            imageVector = Icons.Default.PhotoAlbum,
            contentDescription = null,
            modifier = Modifier.size(80.dp),
            tint = MaterialTheme.colorScheme.primary
        )

        // App name and tagline
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(
                text = "Photo Organizer",
                style = MaterialTheme.typography.headlineLarge,
                color = MaterialTheme.colorScheme.onBackground,
                textAlign = TextAlign.Center
            )
            Text(
                text = "Automatically organize your photos with AI",
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center
            )
        }

        Spacer(modifier = Modifier.height(8.dp))

        // Value proposition bullets
        Column(
            modifier = Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            ValueBullet("Smart AI categorization")
            ValueBullet("Automatic photo sorting")
            ValueBullet("Works offline - your photos stay private")
        }

        Spacer(modifier = Modifier.height(8.dp))

        // Permission card
        PermissionCard(
            title = "Access Your Photos",
            description = "Photo Organizer needs access to your Pictures folder to organize your photos. " +
                    "Your files stay on your device - we never upload them.",
            buttonText = if (isLoading) "Processing..." else "Grant Access",
            onGrantClick = onGrantClick,
            isEnabled = !isLoading
        )

        if (isLoading) {
            CircularProgressIndicator(
                modifier = Modifier.size(24.dp),
                strokeWidth = 2.dp
            )
        }
    }
}

/**
 * Value bullet point component.
 */
@Composable
private fun ValueBullet(text: String) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Icon(
            imageVector = Icons.Default.CheckCircle,
            contentDescription = null,
            modifier = Modifier.size(20.dp),
            tint = MaterialTheme.colorScheme.primary
        )
        Text(
            text = text,
            style = MaterialTheme.typography.bodyLarge,
            color = MaterialTheme.colorScheme.onSurface
        )
    }
}

/**
 * Content shown during folder discovery.
 */
@Composable
private fun DiscoveryContent(
    discoveredFolders: Int,
    totalPhotos: Int,
    isIndeterminate: Boolean
) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(32.dp)
    ) {
        FolderDiscoveryProgress(
            discoveredCount = discoveredFolders,
            totalCount = null, // Indeterminate progress
            photoCount = totalPhotos,
            isIndeterminate = isIndeterminate
        )
    }
}

/**
 * Content shown when onboarding completes successfully.
 */
@Composable
private fun CompleteContent(
    discoveredFolders: Int,
    totalPhotos: Int
) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(24.dp)
    ) {
        Icon(
            imageVector = Icons.Default.CheckCircle,
            contentDescription = null,
            modifier = Modifier.size(80.dp),
            tint = MaterialTheme.colorScheme.primary
        )

        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(
                text = "All Set!",
                style = MaterialTheme.typography.headlineLarge,
                color = MaterialTheme.colorScheme.onBackground
            )
            Text(
                text = "Found $discoveredFolders folders with $totalPhotos photos",
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }

        Text(
            text = "Taking you to your photos...",
            style = MaterialTheme.typography.bodyLarge,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )

        CircularProgressIndicator()
    }
}

/**
 * Content shown when an error occurs.
 */
@Composable
private fun ErrorContent(
    errorMessage: String,
    onRetry: () -> Unit
) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(24.dp)
    ) {
        Icon(
            imageVector = Icons.Default.Error,
            contentDescription = null,
            modifier = Modifier.size(64.dp),
            tint = MaterialTheme.colorScheme.error
        )

        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(
                text = "Something Went Wrong",
                style = MaterialTheme.typography.headlineSmall,
                color = MaterialTheme.colorScheme.onBackground
            )
            Text(
                text = errorMessage,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center
            )
        }

        Button(onClick = onRetry) {
            Text("Try Again")
        }

        TextButton(onClick = onRetry) {
            Text("Skip for now")
        }
    }
}

@Preview(showBackground = true, device = "id:pixel_5")
@Composable
private fun OnboardingScreenPreview() {
    MaterialTheme {
        WelcomeContent(
            isLoading = false,
            onGrantClick = {}
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun DiscoveryContentPreview() {
    MaterialTheme {
        DiscoveryContent(
            discoveredFolders = 5,
            totalPhotos = 127,
            isIndeterminate = true
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun CompleteContentPreview() {
    MaterialTheme {
        CompleteContent(
            discoveredFolders = 8,
            totalPhotos = 245
        )
    }
}

@Preview(showBackground = true)
@Composable
private fun ErrorContentPreview() {
    MaterialTheme {
        ErrorContent(
            errorMessage = "Could not access your photos. Please check your permissions and try again.",
            onRetry = {}
        )
    }
}